lib/galaxy/webapps/galaxy/services/history_contents.py

```diff
@@ -1602,11 +1602,13 @@ def _delete(self, item: HistoryItemModel):
 
     def _undelete(self, item: HistoryItemModel):
         if getattr(item, "purged", False):
-            return
+            raise exceptions.ItemDeletionException("This item has been permanently deleted and cannot be recovered.")
         manager = self._get_item_manager(item)
         manager.undelete(item, flush=self.flush)
 
     def _purge(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
+        if getattr(item, "purged", False):
+            raise exceptions.ItemDeletionException("This item was already permanently deleted.")
```

Comment:
Maybe this can just be a no-op?
```suggestion
            return
```

Reply:
This is intentional, although it might be a better solution, see the *note* in the PR description. Your suggestion was the initial approach, but there is a corner case...
If we select only purged datasets and then try to purge them in bulk (for whatever reason) if we no-op, the API call will return success for all of them in the response, and then the UI will expect a history update with the updated items... which will never occur and we get the `Processing operation...` display forever or until page refresh...

So in summary, if we no-op it's hard to know if we should wait for history changes or not, this way since we know there will be no change, we can drop the processing message at the cost of displaying an error which is kind of true :sweat_smile:

Reply:
Sounds like we need to track the progress of a batch operation, it seems unfortunate that we have to raise an exception here to work around not having a way to do this currently. Could you just touch the HDA update_time for now ? That should also satisfy the update_time filter

Reply:
> Sounds like we need to track the progress of a batch operation

Yes, that would be the ideal solution.

> Could you just touch the HDA update_time for now ? That should also satisfy the update_time filter

Sure, that sounds more reasonable, thanks for the suggestion, will do!

Reply:
By the way, is it ok to raise the error when trying to `undelete` a `purged` dataset or should we do the same, no-op, and just touch the `update_time`?

Reply:
No, I'd raise an exception there.

Reply:
Done! Thanks for the reviews!